kubernetes-sigs/aws-load-balancer-controller
# Fetch VPC ID from runtime using VPC tags provided via controller flags

### Issue

This PR will address the issue raised: #3644
 
### Description

This PR introduces a new runtime argument, aws-vpc-tags which is used to identify VPCs from the list of tags from which we can infer the VPC ID. There is also an optional argument called aws-vpc-name-tag-key incase if the "Name" will not be key-name for the VPC name in tag. 

This is useful for cases where access to AWS metadata is blocked and the VPC ID is not known at deploy time, options to infer the VPC ID are limited


Manual Tests Details

```

Dev Build

➜  ~ kubectl get deploy aws-load-balancer-controller  -n aws-lb-controller -o=jsonpath="{.spec.template.spec.containers[0].args}" 
["--cluster-name=j119106042","--ingress-class=alb","--aws-vpc-tags=Name=j119106042"]%                                                                                                                                                                ➜  ~ 

When I have added the log to display the vpc_id, it got properly fetched the VPC(hidden) based on the given flag --aws-vpc-tags=Name="". 

➜  ~ kubectl logs aws-load-balancer-controller-566f9f556b-dhlbr -n aws-lb-controller 
{"level":"info","ts":"2024-05-09T20:12:03Z","msg":"version","GitVersion":"","GitCommit":"","BuildDate":""}
VPC ID:  vpc-XXXXXXXXXXXX
{"level":"info","ts":"2024-05-09

Current Build

➜  ~ kubectl logs aws-load-balancer-controller-6776c547-7mfj4  -n aws-lb-controller 
{"level":"info","ts":"2024-05-14T10:29:47Z","msg":"version","GitVersion":"","GitCommit":"","BuildDate":""}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"setup","msg":"adding health check for controller"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"setup","msg":"adding readiness check for webhook"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/mutate-v1-pod"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/mutate-v1-service"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/validate-elbv2-k8s-aws-v1beta1-ingressclassparams"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/mutate-elbv2-k8s-aws-v1beta1-targetgroupbinding"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/validate-elbv2-k8s-aws-v1beta1-targetgroupbinding"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/validate-networking-v1-ingress"}
{"level":"info","ts":"2024-05-14T10:29:47Z","logger":"setup","msg":"starting podInfo repo"}
{"level":"info","ts":"2024-05-14T10:29:49Z","logger":"controller-runtime.webhook.webhooks","msg":"Starting webhook server"}
{"level":"info","ts":"2024-05-14T10:29:49Z","msg":"Starting server","kind":"health probe","addr":"0.0.0.0:61779"}
{"level":"info","ts":"2024-05-14T10:29:49Z","msg":"Starting server","path":"/metrics","kind":"metrics","addr":"0.0.0.0:8080"}
{"level":"info","ts":"2024-05-14T10:29:49Z","logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
{"level":"info","ts":"2024-05-14T10:29:49Z","logger":"controller-runtime.certwatcher","msg":"Starting certificate watcher"}
{"level":"info","ts":"2024-05-14T10:29:49Z","logger":"controller-runtime.webhook","msg":"Serving webhook server","host":"","port":9443}

```

### Checklist

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
